corehq/apps/es/client.py

```diff
@@ -0,0 +1,933 @@
+
+        :param index: ``str`` index for which to change the replicas
+        :param replicas: ``int`` number of replicas"""
+        self._validate_single_index(index)
+        settings = {"index.number_of_replicas": replicas}
+        self._index_put_settings(index, settings)
+
+    def index_configure_for_reindex(self, index):
+        """Update an index with settings optimized for reindexing.
+
+        :param index: ``str`` index for which to change the settings"""
+        self._validate_single_index(index)
+        return self._index_put_settings(index, INDEX_CONF_REINDEX)
+
+    def index_configure_for_standard_ops(self, index):
+        """Update an index with settings optimized standard HQ performance.
+
+        :param index: ``str`` index for which to change the settings"""
+        return self._index_put_settings(index, INDEX_CONF_STANDARD)
+
+    def _index_put_settings(self, index, settings):
+        self._validate_single_index(index)
+        if not (list(settings) == ["index"]
+                or all(key.startswith("index.") for key in settings)):
+            raise ValueError(f"Invalid index settings: {settings}")
+        return self._es.indices.put_settings(settings, index)
+
+    def index_put_mapping(self, index, type_, mapping):
+        """Update the mapping for a doc type on an index.
+
+        :param index: ``str`` index where the mapping should be updated
+        :param type_: ``str`` doc type to update on the index
+        :param mapping: ``dict`` mapping for the provided doc type"""
+        self._validate_single_index(index)
+        return self._es.indices.put_mapping(type_, {type_: mapping}, index,
+                                            expand_wildcards="none")
+
+    @staticmethod
+    def _validate_single_index(index):
+        """Verify that the provided index is a valid, single index
+
+        - is non-zero
+        - is not ``_all``
+        - does not contain commas (``,``)
+        - does not contain wildcards (i.e. asterisks, ``*``).
+
+        See <https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-get-index.html#get-index-api-path-params>  # noqa: E501
+
+        :param index: index name or alias"""
+        if not index:
+            raise ValueError(f"invalid index: {index}")
+        elif index == "_all":
+            raise ValueError("refusing to operate on all indices")
+        elif "," in index:
+            raise ValueError(f"refusing to operate on multiple indices: {index}")
+        elif "*" in index:
+            raise ValueError(f"refusing to operate with index wildcards: {index}")
+
+    def reindex(self):
+        """Initiate a reindex operation"""
+        raise NotImplementedError("TODO")
+
+
+class ElasticDocumentAdapter(ElasticClientAdapter):
+    """Base for subclassing document-specific adapters."""
+
+    index_key = None  # set by subclass
+    type = None  # set by subclass
+    mapping = None  # set by subclass
```

Comment:
Rather than setting these to `None` (which is not a valid value), what do you think of adding a note in the docstring that these must be set, and not actually set them here on this base class.